File: sdk/node/src/endpoint.js
/**
 * Where the API is, and how it was decided.
 *
 * One function, so nothing else has to reimplement the discovery order. Two
 * orders that drift is a support question nobody can answer from the stack
 * trace.
 */

/** What `zygo api` listens on when nothing says otherwise. */
export const DEFAULT_URL = 'http://127.0.0.1:7700';

/**
 * Work out which API to talk to.
 *
 * In order: the argument, then `ZYGO_API_URL`, then loopback on the port
 * `zygo api` uses by default. Accepts `unix:///path/to.sock`,
 * `http://host:port`, `https://host:port` and a bare `host:port`.
 *
 * @param {string} [url]
 * @returns {{url: string, socketPath?: string, host: string, port: number, tls: boolean, isUnix: boolean}}
 */
export function resolve(url) {
  return parse(url || process.env.ZYGO_API_URL || DEFAULT_URL);
}

/** @param {string} text */
export function parse(text) {
  const trimmed = String(text).trim();

  if (trimmed.startsWith('unix://')) {
    const socketPath = trimmed.slice('unix://'.length);
    if (!socketPath) {
      throw new TypeError('unix:// needs a path, e.g. unix:///run/user/1000/zygo/api.sock');
    }
    return { url: trimmed, socketPath, host: 'localhost', port: 0, tls: false, isUnix: true };
  }

  let tls = false;
  let rest = trimmed;
  if (trimmed.startsWith('https://')) {
    tls = true;
    rest = trimmed.slice('https://'.length);
  } else if (trimmed.startsWith('http://')) {
    rest = trimmed.slice('http://'.length);
  } else if (trimmed.includes('://')) {
    const scheme = trimmed.split('://', 1)[0];
    throw new TypeError(`\`${scheme}://\` is not an address Zygo serves; use http://, https:// or unix://`);
  }

  // A trailing path is dropped rather than honoured: every route this client
  // calls is rooted, and silently prefixing them would turn a typo in the
  // address into a 404 on every call.
  rest = rest.split('/')[0];
  const colon = rest.lastIndexOf(':');
  const host = colon === -1 ? rest : rest.slice(0, colon);
  const portText = colon === -1 ? (tls ? '443' : '7700') : rest.slice(colon + 1);
  const port = Number(portText);
  if (!Number.isInteger(port) || port <= 0 || port > 65535) {
    throw new TypeError(`\`${portText}\` is not a port number, in \`${text}\``);
  }
  if (!host) throw new TypeError(`\`${text}\` names no host`);

  return {
    url: `${tls ? 'https' : 'http'}://${host}:${port}`,
    host,
    port,
    tls,
    isUnix: false,
  };
}

File: sdk/python/src/zygo/_endpoint.py
# SPDX-License-Identifier: Apache-2.0
"""Where the API is, and how it was decided.

One function, because the rule has to be the same for the synchronous and the
asynchronous client. Two discovery orders that drift is a support question
nobody can answer from the traceback.
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from typing import Optional

#: What ``zygo api`` listens on when nothing says otherwise.
DEFAULT_URL = "http://127.0.0.1:7700"


@dataclass(frozen=True)
class Endpoint:
    """A parsed address: either a unix socket path or a host and port."""

    url: str
    socket_path: Optional[str] = None
    host: str = "127.0.0.1"
    port: int = 7700
    tls: bool = False

    @property
    def is_unix(self) -> bool:
        return self.socket_path is not None

    def __str__(self) -> str:
        return self.url


def resolve(url: Optional[str] = None) -> Endpoint:
    """Work out which API to talk to.

    In order: the argument, then ``ZYGO_API_URL``, then loopback on the port
    ``zygo api`` uses by default. Accepts ``unix:///path/to.sock``,
    ``http://host:port``, ``https://host:port`` and a bare ``host:port``.
    """
    text = url or os.environ.get("ZYGO_API_URL") or DEFAULT_URL
    return parse(text)


def parse(text: str) -> Endpoint:
    text = text.strip()
    if text.startswith("unix://"):
        path = text[len("unix://") :]
        if not path:
            raise ValueError("unix:// needs a path, e.g. unix:///run/user/1000/zygo/api.sock")
        return Endpoint(url=text, socket_path=path)

    tls = False
    rest = text
    if text.startswith("https://"):
        tls, rest = True, text[len("https://") :]
    elif text.startswith("http://"):
        rest = text[len("http://") :]
    elif "://" in text:
        scheme = text.split("://", 1)[0]
        raise ValueError(f"`{scheme}://` is not an address Zygo serves; use http://, https:// or unix://")

    # A trailing path is dropped rather than honoured: every route this client
    # calls is rooted, and silently prefixing them would turn a typo in the
    # address into a 404 on every call.
    rest = rest.split("/", 1)[0]
    host, _, port_text = rest.rpartition(":")
    if not host:
        host, port_text = rest, "443" if tls else "7700"
    try:
        port = int(port_text)
    except ValueError as e:
        raise ValueError(f"`{port_text}` is not a port number, in `{text}`") from e

    scheme = "https" if tls else "http"
    return Endpoint(url=f"{scheme}://{host}:{port}", host=host, port=port, tls=tls)
